mlir/include/mlir/Interfaces/SideEffectInterfaceBase.td

```diff
@@ -183,6 +184,9 @@ class SideEffect<EffectOpInterfaceBase interface, string effectName,
 
   // Does this side effect act on every single value of resource.
   bit effectOnFullRegion = range.Value;
+
+  // Does this side effect potentially occur after op exit
+  bit asynchronous = isAsync;
```

Comment:
This is a tricky semantics to add. I actually am not sure how to model this.
How are we supposed to reason about Effects which have this bit set?

Feel this is deep enough of a change that we should have a clear description of this in a proposal on Discourse.

Reply:
That's fair. I will mark this as a draft until I can work out the proposal with my teammates.

Reply:
FYI, effect instances have an attribute field for additional qualification of effects. It may be worth using, rather than pushing `asynchronous` on every potential client, as long as one can come up with a "conservatively safe" way of handling side effects while ignoring this information.

Reply:
I have a variation that fills the attribute parameter with a BoolAttr. How would you recommend making it certain who added the parameter and for what purpose?

Reply:
Using a custom attribute (GpuAsyncAttr) that stores a `bool` (or even make it unit if that makes sense) rather than a `BoolAttr`. There was some previous patch where we discussed putting a `DictAttr` and using string keys.

Note that I'm not sure this is the right mechanism, it's conditioned on there being a way to interpret the absence of the attribute in a conservatively correct way. (The only user of the attribute I'm aware of so far is a downstream project that uses affine-ish descriptors of the subset of memref elements accessed, in absence of which analyses can just conservatively assume all elements are.)